sapicore/engine/component/__init__.py

```diff
@@ -0,0 +1,183 @@
+        if configuration.get("loggable"):
+            self._loggable_props_ = configuration.get("loggable")
+
+        # if `log_destination` was passed, save the configuration dictionary to that location.
+        if os.path.exists(log_destination):
+            save_yaml(self.configuration, log_destination)
+
+    def forward(self, data: Tensor) -> dict:
+        """Processes an input, updates the state of this component, and advances the simulation by one step.
+
+        Parameters
+        ----------
+        data: Tensor
+            Input to be processed (e.g., added to a neuron's numeric state tensor `voltage`).
+
+        Returns
+        -------
+        dict
+            A dictionary whose keys are loggable attributes and whose values are their states as of this time step.
+            For potential use by a :class:`~pipeline.simulation.Simulator` or any other :class:`~pipeline.Pipeline`
+            script handling runtime operations.
+
+        Raises
+        ------
+        NotImplementedError
+            The forward method must be implemented by each derived class.
+
+        """
+        raise NotImplementedError
+
+    def heterogenize(self, unravel: bool = True):
+        """Edits configurable tensor attributes based on a sweep search dictionary if one was provided by the user
+        within this object's `configuration` dictionary.
+
+        Note
+        ----
+        If a `sweep` key is not present in `configuration`, this method will pass silently, retaining the existing
+        configurable values. It can be invoked safely from generic methods (e.g.,
+        :meth:`~engine.network.Network.build`).
+
+        Parameters
+        ----------
+        unravel: bool
+            When the property is a 2D tensor, dictates whether combination values should be assigned
+            to individual elements (True) or overwrite entire rows (False).
+
+            The former behavior is appropriate for synapse attributes. It is achieved using `numpy.unravel_index`,
+            which maps flattened-like indices to their nD equivalents given the shape of the array.
+
+            The latter behavior is called for when values are themselves 1D vectors (e.g., oscillator frequencies).
+
+        Warning
+        -------
+        This method mutates the underlying component object. To simply view the combinations, initialize a
+        :class:`~utils.sweep.Sweep` object with your desired search space dictionary and number of combinations.
+
+        """
+        # safe to call if object has no sweep specification or empty configuration; method will return silently.
+        if self.configuration and self.configuration.get("model", {}).get("sweep"):
+            # the sweep object maintains a list of valid parameter combinations, each of which is a dictionary.
+            sweep = Sweep(
+                search_space=self.configuration.get("model", {}).get("sweep"), num_combinations=self.num_units
```

Comment:
Great catch. It slipped under my radar because it's only been used with derivative classes that have `num_units` so far, but that doesn't have to be the case. I'll pass `num_combinations` as an argument instead.